### PR TITLE
[ConstraintSystem] Implement simplification for all locator elements

### PR DESCRIPTION
**Pull request by xedin**
Remove the `default` in `simplifyLocatorToAnchor` and implement simplification
logic for all missing elements. This makes `simplifyLocatorToAnchor` future-proof.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility debug

**Review by hborla — APPROVED** [review]
Nice!

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -566,6 +566,13 @@ template <typename T = Decl> T *getAsDecl(ASTNode node) {
   return nullptr;
 }
 
+template <typename T = Decl>
```
Comment (LucianoPAlmeida) [inline include/swift/Sema/ConstraintSystem.h:569]: Just curious, IIUC wouldn't default template parameter be <typename T = Stmt>?

Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h:569]: Ah, that’s right!

Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h:569]: So much for copying… 🤦‍♂️  I’ll wait until source compact is done and amend the corresponding commit.

**Comment by xedin** [comment]
Failure in swift-protobuf-plugin is unrelated - `sandbox-exec: <internal init prelude>:456:12: unable to open system.sb: not found`

File: include/swift/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h]: Good catch, fixed!

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**